include/quicr/message_buffer.h

```diff
@@ -106,87 +104,145 @@ class MessageBuffer
+   * @brief Overload for Namespace integral type to write in NBO.
+   * @param msg The MessageBuffer to write to.
+   * @param value The message to be written.
+   * @returns The MessageBuffer that was written to.
+   */
+  friend MessageBuffer& operator<<(MessageBuffer& msg, quicr::Namespace value)
+  {
+    if constexpr (std::endian::native == std::endian::big)
+      return msg << value.name() << value.length();
 
-  return ((value >> 8) & 0x00ff) | ((value << 8) & 0xff00);
-}
+    return msg << value.length() << value.name();
+  }
 
-constexpr uint32_t
-swap_bytes(uint32_t value)
-{
-  if constexpr (std::endian::native == std::endian::big)
-    return value;
+  /**
+   * @brief A fancy operator for pop, reads a byte off the buffer.
+   * @param buffer The MessageBuffer to read from.
+   * @param value The value to read into.
+   * @returns The MessageBuffer that was read from.
+   */
+  friend inline MessageBuffer& operator>>(MessageBuffer& msg, uint8_t& value)
+  {
+    if (msg.empty()) {
+      throw MessageBuffer::ReadException(
+        "Cannot read from empty message buffer");
+    }
+
+    value = msg.front();
+    msg.pop();
+    return msg;
+  }
 
-  return ((value >> 24) & 0x000000ff) |
-         ((value >>  8) & 0x0000ff00) |
-         ((value <<  8) & 0x00ff0000) |
-         ((value << 24) & 0xff000000);
-}
+  /**
+   * @brief Reads QUICR integral types in HBO.
+   * @tparam T A type satisfying quicr::is_integral.
+   * @param buffer The MessageBuffer to read from.
+   * @param value The value to read into.
+   * @returns The MessageBuffer that was read from.
+   */
+  template<typename T, typename = std::enable_if_t<quicr::is_integral_v<T>, T>>
+  friend inline MessageBuffer& operator>>(MessageBuffer& msg, T& value)
+  {
+    if (msg.empty()) {
+      throw MessageBuffer::ReadException(
+        "Cannot read from empty message buffer");
+    }
 
-constexpr uint64_t
-swap_bytes(uint64_t value)
-{
-  if constexpr (std::endian::native == std::endian::big)
-    return value;
+    if (msg._buffer.size() < sizeof(T)) {
+      throw MessageBuffer::ReadException(
+        "Cannot read mismatched size buffer into size of type: Wanted " +
+        std::to_string(sizeof(T)) + " but buffer only contains " +
+        std::to_string(msg._buffer.size()));
+    }
 
-  return ((value >> 56) & 0x00000000000000ff) |
-         ((value >> 40) & 0x000000000000ff00) |
-         ((value >> 24) & 0x0000000000ff0000) |
-         ((value >>  8) & 0x00000000ff000000) |
-         ((value <<  8) & 0x000000ff00000000) |
-         ((value << 24) & 0x0000ff0000000000) |
-         ((value << 40) & 0x00ff000000000000) |
-         ((value << 56) & 0xff00000000000000);
-}
+    auto val_ptr = reinterpret_cast<uint8_t*>(&value);
+    std::memcpy(val_ptr, msg._buffer.data(), sizeof(T));
 
-constexpr quicr::Name
-swap_bytes(quicr::Name value)
-{
-  if constexpr (std::endian::native == std::endian::big)
-    return value;
+    msg._buffer.erase(msg._buffer.begin(),
```

Comment:
I suppose this is one reason to have exceptions in these friend classes, though it would be nice if one could just call `ReadValue()` in this case and have the exception thrown in the MessageBuffer.  This is what Cantina's DataBuffer does.  It lacks these operators, but it addresses this sort of issue.  Also, doing this erase() is truly unfortunate, as it causes memmove() repeatedly.  In Cantina's DataBuffer, we keep references to the current read position, as opposed to moving memory.

Reply:
We could absolutely do that. We did that in time_queue, could implement it here easily as well.